lib/manageiq/providers/openstack/legacy/events/openstack_stf_event_monitor.rb

```diff
@@ -0,0 +1,112 @@
+require 'manageiq/providers/openstack/legacy/openstack_event_monitor'
+require 'manageiq/providers/openstack/legacy/events/openstack_event'
+require 'manageiq/providers/openstack/legacy/events/openstack_stf_event_converter'
+require 'manageiq/providers/openstack/legacy/events/openstack_stf_event_receiver'
+require 'manageiq/providers/openstack/legacy/events/openstack_stf_event_test_receiver'
+require 'qpid_proton'
+
+class OpenstackStfEventMonitor < OpenstackEventMonitor
+  DEFAULT_AMQP_PORT  = 5666
+  DEFAULT_TOPIC_NAME = 'anycast/ceilometer/event.sample'.freeze
+
+  def self.available?(options = {})
+    $log.info("Testing connection to STF..") if $log
+    $log.debug("With STF options: #{options.inspect}") if $log
+    qdr_client = Qpid::Proton::Container.new(OpenStackStfEventTestReceiver.new(build_qdr_client_url(options), DEFAULT_TOPIC_NAME))
+    qdr_client.run
+    true
+  rescue => ex
+    $log.info("OpenstackSTFEventMonitor availability check failed with #{ex}.") if $log
+    false
+  ensure
+    qdr_client.close if qdr_client.respond_to?(:close)
+    qdr_client.stop if qdr_client.respond_to?(:stop)
+  end
+
+  def self.plugin_priority
+    1
+  end
+
+  def self.build_qdr_client_url(options)
+    $log.debug("Building STF QDR client connection with #{options.inspect}..") if $log
+    protocol = options[:security_protocol].to_s.start_with?('ssl') ? 'amqps' : 'amqp'
+    hostname = options[:hostname]
+    port     = options[:port] || DEFAULT_AMQP_PORT
+    "#{protocol}://#{hostname}:#{port}"
+  end
+
+  def initialize(options = {})
+    $log.info("Building STF QDR client INIT #{options.inspect}..") if $log
+    @options = options
+    @ems = options[:ems]
+    @config = options.fetch(:stf, {})
+
+    @events = []
+    @events_mutex = Mutex.new
+
+    @recv_block = ->(event) { @events << event }
+
+    @qdr_receiver = Qpid::Proton::Container.new(OpenStackStfEventReceiver.new(self.class.build_qdr_client_url(@options), @config[:topic_name] || DEFAULT_TOPIC_NAME, @recv_block, @events_mutex))
+  end
+
+  def start
+    return if @qdr_receiver.running > 0
+
+    $log.info("STF QDR client START..") if $log
+    @handler = Thread.start { @qdr_receiver.run }
+  end
+
+  def stop
+    $log.info("STF QDR client STOP..") if $log
+    @qdr_receiver&.stop
+    @handler&.terminate!
+  end
+
+  def each_batch
+    @collecting_events = true
+    while @collecting_events
+      @events_mutex.synchronize do
```

Comment:
Every time I see a mutex I wonder if we could do this without a mutex haha but we can look at that in a followup

Reply:
Okay :-) Do you mean using other thread data sharing construct or refactor the code to way that the sync would not be needed?

Reply:
Using an already concurrent safe [queue](https://ruby-doc.org/core-2.5.0/Queue.html) would allow us to push from the collector thread and consume from the main thread without implementing our own locking

Reply:
Thanks for suggestion, updated by https://github.com/ManageIQ/manageiq-providers-openstack/pull/589